### ISSUE TITLE
Support retrieving original type name from mangled type name

### ISSUE THREAD

**Issue by captainsafia** [dotnet/roslyn#55651]
## Background and Motivation

For https://github.com/dotnet/aspnetcore/issues/34540, we wanted to add a feature that would automatically set some internal metadata to the name

This presented a bit of a problem for local functions. For a method defined as follows:

```
var app = WebApplication.Create(args);
string Greeting() => "Hello, World!";
app.MapGet("/hello", Greeting);
```

When the `MapGet` is executed at runtime above, we want to populate some metadata with the following using reflection APIs.

```
EndpointName = "Greeting"
```

However, because `Greeting` is compiler generated, the computed type name is mangled as `<Program>$.<<Main>$>g__Greeting|0_0()`.

Which makes it impossible to produce the desired metadata at runtime.

## Proposed API

```diff
namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.All, Inherited = true)]
    public sealed class CompilerGeneratedAttribute : Attribute
    {
        public CompilerGeneratedAttribute() { }

+      public string OriginalName { get; }
    }
}
```

## Usage Examples

```csharp
var compilerGeneratedName = Attribute.GetCustomAttribute(type, typeof (CompilerGeneratedAttribute));
builder.Metadata.Add(new EndpointNameMetadata(compilerGeneratedName.OriginalName));
```

## Alternative Designs

During an out-of-band conversation, the following alternatives were proposed.

- Avoid mangling type names all together for compiler generated types.
- Expose an API that allows end-users to derive how a mangled type is produced given an original name. Allows users to make the mangled name -> original name mapping themselves.

**Comment by tmat** [comment]
I don't think we should emit another attribute with information that is already in the metadata. That would just increase the size of the generated assemblies.

Instead Roslyn can publish a source NuGet package that provides APIs for parsing mangled names.
This package would contain [GeneratedNameParser](https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNameParser.cs) and ac couple of dependent types.

**Comment by captainsafia** [comment]
> I don't think we should emit attribute with information that is already in the metadata. That would just increase the size of the generated assemblies.

FWIW, this is proposing amending the existing `CompilerGeneratedAttribute` as opposed to producing a new attribute altogether.

> This package can be referenced by any library that needs to parse these names. There are other scenarios where mangled names need to be parsed - e.g. stack trace prettyfier.

Good to know. I added an "Other Scenarios" section to compile some of these. Feel free to add any others you know.

**Comment by tmat** [comment]
The "other scenarios" might need more information than just original name. E.g. for lambdas and local functions the name of the containing method is useful when displaying stack trace info. We could add another property to `CompilerGeneratedAttribute` attribute, which would also duplicate the info already encoded in the mangled name. Seems like instead of bloating the metadata we can just enable reading of the existing data.

**Comment by Youssef1313** [comment]
Would CallerArgumentExpression help here?

**Comment by captainsafia** [comment]
> Would CallerArgumentExpression help here? ie, change the signature to:

Yep -- this would work, but it would require us modifying a public API and funneling the `endpointName` through several layers of logic so it's not ideal (great recommendation none the less).

For now, we've relied on the strategy of incorporating some parsing utils in our codebase as a temporary stopgap until we revisit this for .NET 7.
